shard.go

```diff
@@ -0,0 +1,93 @@
+package wavelet
+
+import (
+	"sync"
+)
+
+const (
+	shardsCount = 64
+)
+
+// TransactionMap is a sharded map with TransactionID as its key
+type TransactionMap [shardsCount]*transactionMapShard
+
+type transactionMapShard struct {
+	sync.Mutex
+
+	items map[TransactionID]interface{}
+}
+
+func NewTransactionMap() TransactionMap {
+	var shards TransactionMap
+	for i := uint64(0); i < shardsCount; i++ {
+		shards[i] = &transactionMapShard{
+			items: make(map[TransactionID]interface{}, 2048),
+		}
+	}
+
+	return shards
+}
+
+func (c TransactionMap) Len() uint64 {
+	var l uint64
+
+	for i := uint64(0); i < shardsCount; i++ {
+		c[i].Lock()
+		l += uint64(len(c[i].items))
+		c[i].Unlock()
+	}
+
+	return l
+}
+
+func (c TransactionMap) Iterate(fn func(key TransactionID, value interface{}) bool) {
+	for i := uint64(0); i < shardsCount; i++ {
+		c[i].Lock()
+
+		for key, value := range c[i].items {
+			if !fn(key, value) {
+				c[i].Unlock()
+				return
+			}
+		}
+
+		c[i].Unlock()
+	}
+}
+
+func (c TransactionMap) Get(key TransactionID) (interface{}, bool) {
+	shard := c.GetShard(key)
+
+	shard.Lock()
+	defer shard.Unlock()
+
+	if tx, exists := shard.items[key]; !exists {
+		return nil, false
+	} else {
+		return tx, true
+	}
+}
+
+func (c TransactionMap) Put(key TransactionID, value interface{}) {
+	shard := c.GetShard(key)
+
+	shard.Lock()
+	defer shard.Unlock()
+
+	shard.items[key] = value
+}
+
+func (c TransactionMap) Delete(key TransactionID) {
+	shard := c.GetShard(key)
+
+	shard.Lock()
+	defer shard.Unlock()
+
+	delete(shard.items, key)
+}
+
+func (c TransactionMap) GetShard(key TransactionID) (shard *transactionMapShard) {
+	// Use the first byte as the shard ID
+	shardID := uint64(key[0]) % shardsCount
```

Comment:
Rather than use modulus, could you try it with this and re-run benchmarks?

`(uint64(key[0]) * uint64(shardsCount)) >> 32 `

Reply:
Inspiration: https://lemire.me/blog/2016/06/27/a-fast-alternative-to-the-modulo-reduction/

This assumes that `shardsCount` is a power of 2 btw.

Reply:
It's slower:

![2019-12-05_13-29](https://user-images.githubusercontent.com/1431045/70206677-4381c280-1763-11ea-8473-9faf75c7e724.png)

Reply:
That seems ... strange; what about `uint64(key[0]) & uint64(shardsCount - 1)`? Also, since we can fit shards into a uint16, we can make the keys uint16 and do `uint16(key[0] & uint16(shardsCount - 1)`.

Reply:
About the same as with modulo:
![2019-12-05_13-53](https://user-images.githubusercontent.com/1431045/70207745-9a3ccb80-1766-11ea-9949-fd57060542d2.png)

Reply:
Okay - FYI, since you are just using key[0], you can limit to just using 256 shards.